Support "control after generate" in vue

AustinMroz opened the pull request:
Continuation of #6034 with
- Updated synchronization for seed
- Properly truncates the displayed widget value for the button
- Synchronizes control after generate state with litegraph and allows for serialization

Several issues from original PR have not (yet) been addressed, but are likely better moved to future PR
- fix step value being 10 (legacy system)
- ensure it works with COMBO (Fixed in #7095)
- ensure it works with FLOAT (Fixed in #7095)
- either implement or remove the config button functionality - think it should open settings?

<img width="280" height="694" alt="image" src="https://github.com/user-attachments/assets/f36f1cb0-237d-4bfc-bff1-e4976775cf98" />

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-6985-Support-control-after-generate-in-vue-2b86d73d365081d8b01ce489d887ff00) by [Unito](https://www.unito.io)

christian-byrne committed:
add control_after_generate ui

christian-byrne committed:
don't use forEach

christian-byrne committed:
seed widget

christian-byrne committed:
seed widget2

christian-byrne committed:
handle legacy step value

christian-byrne committed:
feat: update NumberControlPopover with semantic design tokens

christian-byrne committed:
fix test

AustinMroz committed:
Fix display of controled widget values

AustinMroz committed:
Revert useGraphNodeManager changes

AustinMroz committed:
Merge origin/main

AustinMroz committed:
Swap NumberInputs to rekka ui to embed control

The @pointerup.stop was breaking reka ui NumberFields.

IIRC, this was added to allow selecting text without dragging nodes.
Current testing suggests this isn't required for pointerup

This reduces the margins some on number inputs. It's trivial to add a
px-2.5, but helps with information density

AustinMroz committed:
Persist control widget value

Commit:
[automated] Update test expectations

AustinMroz committed:
Test fixes and nits

lostdisc commented:
>either implement or remove the config button functionality - think it should open settings?

Personally as a random SDXL user, I do think the "control _before_ generate" setting should be made a lot more obvious (or better yet, the default), since it's more intuitive for users wanting to quickly retry a seed.  In before-mode, when I get a promising-but-not-perfect result from a random seed, I can just switch to "fixed" mode, tweak my prompt, and rerun.

In after-mode, by contrast, the seed has already changed, and I have to retrieve the old value by digging into the image's metadata in a separate program and manually copy-pasting it back into the seed field.  I can't even copy seeds from the job history, only a useless* job ID.

*: I'm sure it's useful for something, just like control-after-generate must be useful for something, but I can't fathom why after-mode is the default.  So I'm glad there's a before/after setting, but I wish its existence and lifesaving practical effect were made more obvious.

christian-byrne commented:
> > either implement or remove the config button functionality - think it should open settings?
> 
> Personally as a random SDXL user, I do think the "control _before_ generate" setting should be made a lot more obvious (or better yet, the default), since it's more intuitive for users wanting to quickly retry a seed. In before-mode, when I get a promising-but-not-perfect result from a random seed, I can just switch to "fixed" mode, tweak my prompt, and rerun.
> 
> In after-mode, by contrast, the seed has already changed, and I have to retrieve the old value by digging into the image's metadata in a separate program and manually copy-pasting it back into the seed field. I can't even copy seeds from the job history, only a useless* job ID.
> 
> *: I'm sure it's useful for something, just like control-after-generate must be useful for something, but I can't fathom why after-mode is the default. So I'm glad there's a before/after setting, but I wish its existence and lifesaving practical effect were made more obvious.

This is great feedback, what would you say would be most intuitive or obvious to you?

lostdisc commented:
I think "control before generate" would be the more intuitive default.  With "control after generate", it's confusing for the seed to change at the end of a gen, since it no longer corresponds with the output on display, and there's no way within ComfyUI to retrieve the old seed for a retry with different prompts/settings, which is a common use case.

The one downside I can think of for "control before generate" is that increment/decrement would also change the starting seed for the first gen, which would be unexpected.  Maybe it can skip changing the seed when the seed value has just been manually changed by the user.  (This exception would also make sense for randomize mode.)

christian-byrne committed:
Merge branch 'main' into austin/vue-control-after-generate

Commit:
[automated] Update test expectations

DrJKL commented:
> I think "control before generate" would be the more intuitive default. With "control after generate", it's confusing for the seed to change at the end of a gen, since it no longer corresponds with the output on display, and there's no way within ComfyUI to retrieve the old seed for a retry with different prompts/settings, which is a common use case.
> 
> The one downside I can think of for "control before generate" is that increment/decrement would also change the starting seed for the first gen, which would be unexpected. Maybe it can skip changing the seed when the seed value has just been manually changed by the user. (This exception would also make sense for randomize mode.)

There is an option in the settings to change it to control before generate.
<img width="680" height="186" alt="image" src="https://github.com/user-attachments/assets/50f28cb2-bf1b-4d84-9faf-e6e129ff2565" />

AustinMroz committed:
Add max and mins from litegraph implementation

Without these, randomized values can exceed the maximum safe integer
range which results in the increment and decrement buttons being removed

lostdisc commented:
> There is an option in the settings to change it to control before generate.

I know; that's how I can articulate the concept now.  But until recently, I was just suffering through the default mode with no idea it could be changed, until someone else pointed out this setting existed and did something useful.  I bet many users follow the same path, or never find out.

I think putting a toggle right in the KSampler node would help a lot with discoverability, but since it's a shared setting, maybe it's hard to keep the UI state synced across multiple nodes?

Barring that, maybe the simplest improvement would be changing the terminology from "control before/after generate" to "**update seed** before/after generate", which makes the practical effect much clearer.  Similarly, the setting could be relabeled from "widget control mode: before/after" to "update seed fields: before generate / after generate".

(Or is the current wording a vague "control" because it also affects things other than seed RNGs?)

AustinMroz committed:
Merge 8e006bb8a306^

AustinMroz committed:
Merge main

Commit:
[automated] Update test expectations

AustinMroz committed:
Empty commit to force tests to rerun

AustinMroz committed:
Fix number precision

Commit:
[automated] Update test expectations

AustinMroz committed:
Merge main

Commit:
[automated] Update test expectations

AustinMroz committed:
Merge main

AustinMroz committed:
Revert to primevue, fix test

Playwright locators break because there's now an additional button

Swapping to reka ui introduces a bug where clicking and dragging the
mouse off a button causes the number to change continually. This is hard
to change and I found a workaround to get truncation to work with
primevue instead.

AustinMroz committed:
Revert vitest changes, clear browser snapshots

Commit:
[automated] Update test expectations

AustinMroz committed:
Empty commit to trigger tests

AustinMroz committed:
Merge main

Commit:
[automated] Update test expectations

AustinMroz committed:
Empty commit to force tests

DrJKL commented:
@coderabbitai full review

AustinMroz committed:
nits

DrJKL commented on src/renderer/extensions/vueNodes/widgets/composables/useStepperControl.ts:
```diff
@@ -0,0 +1,117 @@
+  DECREMENT = 'decrement',
+  RANDOMIZE = 'randomize',
+  LINK_TO_GLOBAL = 'linkToGlobal'
+}
```
```suggestion
export const NumberControlMode = {
  FIXED: 'fixed',
  INCREMENT: 'increment',
  DECREMENT: 'decrement',
  RANDOMIZE: 'randomize',
  LINK_TO_GLOBAL: 'linkToGlobal'
} as const
```

That thing we talked about

DrJKL commented on src/renderer/extensions/vueNodes/widgets/composables/useStepperControl.ts:
```diff
@@ -0,0 +1,117 @@
+
+  // Register with singleton registry
+  onMounted(() => {
+    numberControlRegistry.register(controlId, applyControl)
```
Double check to see if you need to be DOM dependent here.

AustinMroz replied:
Followup: sorta. It's being used as a convenient way to ensure it gets unregistered at some point. Since I'm proposing removal in a future PR, it's probably not worth rewriting the implementation here.

DrJKL commented on src/stores/globalSeedStore.ts:
I think this piece could be pruned for now, it's a deserving feature and should be treated with such deference.

AustinMroz committed:
Further nits

AustinMroz committed:
Merge main

Commit:
[automated] Update test expectations

AustinMroz committed:
Remove global seed tests

AustinMroz committed:
Merge remote-tracking branch 'origin/main' into austin/vue-control-after-generate

Commit:
[automated] Update test expectations

christian-byrne committed:
merge: main into austin/vue-control-after-generate

AustinMroz committed:
Merge branch 'main' into austin/vue-control-after-generate

christian-byrne committed:
fix: restore executeNumberControls calls removed in merge

christian-byrne committed:
fix: remove space-y-1 causing extra margin on number inputs

Commit:
[automated] Update test expectations
